syntax = "proto3";

package ptre;

service Rdma {
  rpc GetLID(GetLIDRequest)
    returns (GetLIDResponse);

  rpc GetQPAttr(GetQPAttrRequest)
    returns (GetQPAttrResponse);

  rpc GetRemoteAddress(GetRemoteAddressRequest)
    returns (GetRemoteAddressResponse);

  rpc GetRemoteParamAddress(GetRemoteParamAddressRequest)
    returns (GetRemoteParamAddressResponse);

  rpc AttemptPush(AttemptPushRequest)
    returns (AttemptPushResponse);

  rpc AttemptPushVar(AttemptPushVarRequest)
    returns (AttemptPushVarResponse);

  rpc NotifyPushDone(NotifyPushDoneRequest)
    returns (NotifyPushDoneResponse);

  rpc Barrier(BarrierRequest)
    returns (BarrierResponse);

  rpc Recv(RecvRequest)
    returns (RecvResponse);

  rpc GetRemoteAddressV2(GetRemoteAddressV2Request)
    returns (GetRemoteAddressV2Response);

  rpc GetPermit(GetPermitRequest)
    returns (GetPermitResponse);
}

message MemoryRegion {
  uint64 remote_addr = 1;
  uint32 rkey = 2;
}

// GetLID
message GetLIDRequest { }
message GetLIDResponse {
  int32 lid = 1;
}

// GetQPAttr
message GetQPAttrRequest {
  int32 src_rank = 1;
}
message GetQPAttrResponse {
  uint32 qpn = 1;
  uint32 psn = 2;
}

// GetRemoteAddress
message GetRemoteAddressRequest {
  BufType buf_type = 1;
  string var_name = 2;
}

message GetRemoteAddressResponse {
  uint64 remote_addr = 1;
  uint32 rkey = 2;
}

message GetRemoteEnvRequest {
  int32 rank = 1;
}

message GetRemoteEnvResponse {
  int32 rank = 1;
  int32 lid = 2;
  int32 qpn = 3;
  uint64 snp = 4;  // global subnet prefix
  uint64 iid = 5;  // global interface id
}

message GetRemoteParamAddressRequest {
  int32 rank = 1;
}

message GetRemoteParamAddressResponse {
  int32 rank = 1;
  repeated MemoryRegion mr = 3;
}

message AttemptPushRequest {
  int32 rank = 1;
  int32 vstep = 2;
}

message AttemptPushResponse {
  bool available = 1;
}

message NotifyPushDoneRequest {
  int32 src_rank = 1;
  string var_name = 2;
}

message NotifyPushDoneResponse {
}

message BarrierRequest {
}
message BarrierResponse {
  bool entered = 1;
}

enum BufType {
  BUF_TYPE_RAW = 0;
  BUF_TYPE_RECV_BUF = 1;
  BUF_TYPE_SEND_BUF = 2;
  BUF_TYPE_AGG_BUF_STATE = 3;
  BUF_TYPE_AGG_BUF = 4;
  BUF_TYPE_FLAG_RECV = 5;
  BUF_TYPE_FLAG_SEND = 6;
  BUF_TYPE_BARRIER_COUNTER = 7;
  BUF_TYPE_BARRIER_RELEASE = 8;
  BUF_TYPE_PUSH_PERMIT = 9;
  BUF_TYPE_PUSH_PERMIT_READ = 10;
}

message GetRemoteAddressV2Request {
  int32 rank = 1;
  BufType type = 2;
  string name = 3;
}

message GetRemoteAddressV2Response {
  int32 rank = 1;
  BufType type = 2;
  string name = 3;
  repeated MemoryRegion mr = 4;
}

message RecvRequest {
  int32 dst_rank = 1;
  uint32 len = 2;
  string name = 3;
}

message RecvResponse {
  bytes buf = 1;
}

message GetPermitRequest {
  string var_name = 1;
}

message GetPermitResponse {
  int32 permit = 1;
}

message AttemptPushVarRequest {
  string var_name = 1;
  int32 src_rank = 2;
}

message AttemptPushVarResponse {
  bool result = 1;
}
